framework_crates/bones_ecs/src/entities.rs

```diff
@@ -351,6 +441,89 @@ impl<'a, I: Iterator> Iterator for EntitiesIterWith<'a, I> {
 }
 
 impl Entities {
+    /// Get a single entity and components in the given query if there is exactly one entity
+    /// matching the query.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if the number of matching entities is not *exactly one*.
+    pub fn single_with<Q: QueryItem>(
+        &self,
+        query: Q,
+    ) -> (Entity, <<Q as QueryItem>::Iter as Iterator>::Item) {
+        self.get_single_with(query).unwrap()
+    }
+
+    /// Get a single entity and components in the given query if there is exactly one entity
+    /// matching the query.
+    pub fn get_single_with<Q: QueryItem>(
+        &self,
+        query: Q,
+    ) -> Result<(Entity, <<Q as QueryItem>::Iter as Iterator>::Item), QuerySingleError> {
+        let mut bitset = self.bitset().clone();
+        query.apply_bitset(&mut bitset);
+
+        let entity = {
```

Comment:
I'm wondering if we could instead of iterating over entities here, apply the entity's bitset in query.get_single_with_bitset(bitset), and only filter by bitset once during that iteration.

Reply:
Without iterating over the entities bitset we wouldn't know the index of the found value to construct the `Entity` to return. We could make `UntypedComponentStore::get_single_with_bitset` return a `(u32, SchemaRef)` instead of just the ref and pass it up the chain. idk is that fine or is it a code smell?

Reply:
Ah gotcha, I understand.

I'm thinking that maybe we can do `query.iter_with_bitset(bitset)` first to get iterator, and then do similar logic you have in which you check if iterator is_some, and that next is_none to confirm is single.

In the case that we know it is single, I'm wondering if we can get the `current_id` off of the iterator, and then lookup the entity id using this current_id (should be index in bitset?). This way we can skip the iterating over entities by relying on its bitset being applied to query item, but hopefully still get entity id for mapping single to `(entity, item)`.

I think this is sound but definitely check me on that if doesn't go well or if ends up being a pain. Definitely ok merging this as is, if we ever profile and find there are perf issues with this (I kind of doubt we will) can rethink it, but not too worried.